src/query_engine/query_element/Terminal.h

```diff
@@ -0,0 +1,207 @@
+
+    /**
+     * Protected constructor.
+     */
+    Terminal() : QueryElement() {
+        this->handle = shared_ptr<char>{};
+        this->is_variable = false;
+        this->is_terminal = true; // overrrides QueryElement default
+    }
+
+public:
+
+    /**
+     * Destructor.
+     */
+    ~Terminal() {};
+
+    /**
+     * Empty implementation. There are no QueryNode element to setup.
+     */
+    void virtual setup_buffers() {}
+
+    /**
+     * Empty implementation. There are no QueryNode element or local thread to shut down.
+     */
+    void virtual graceful_shutdown() {}
+                                
+    /**
+     * Returns a string representation of this Terminal (mainly for debugging; not optimized to
+     * production environment).
+     */
+    virtual string to_string() = 0;
+
+    /**
+     * A flag to indicate whether this Terminal is a Variable or not.
+     */
+    bool is_variable;
+
+    /**
+     * Handle of the terminal.
+     */
+    shared_ptr<char> handle;
+
+    /**
+     * Name of the terminal.
+     *
+     * Actually, only Nodes and Variables have names; Links' name is an empty string.
+     */
+    string name;
+};
+
+// -------------------------------------------------------------------------------------------------
+// Node
+
+/**
+ * QueryElement which represents a node.
+ */
+class Node : public Terminal {
+
+public:
+
+    /**
+     * Constructor.
+     *
+     * @param type Type of the node.
+     * @param name Name of the node.
+     */
+    Node(const string &type, const string &name) : Terminal() {
+        this->type = type;
+        this->name = name;
+        this->handle = shared_ptr<char>(terminal_hash((char *) type.c_str(), (char *) name.c_str()));
+    }
+
+    /**
+     * Returns a string representation of this Node (mainly for debugging; not optimized to
+     * production environment).
+     */
+    string to_string() {
+        return "<" + this->type + ", " + this->name + ", " + string(this->handle.get()) + ">";
+    }
+
+    /**
+     * Type of this node.
+     */
+    string type;
+};
+
+// -------------------------------------------------------------------------------------------------
+// Link
+
+/**
+ * QueryElement which represents a link.
+ */
+template <unsigned int ARITY>
+class Link : public Terminal {
+
+public:
+
+    /**
+     * Constructor.
+     *
+     * @param type Type of the Link.
+     * @params targets Array with targets of the Link. Targets are supposed to be
+     *         handles (i.e. strings). No nesting of Nodes or other Links are allowed.
+     */
+    Link(const string &type, const array<string, ARITY> &targets) : Terminal() {
+        this->name = "";
+        this->type = type;
+        this->targets = targets;
+        this->arity = ARITY;
+        char *handle_keys[ARITY + 1];
+        handle_keys[0] = (char *) named_type_hash(type.c_str());
+        for (unsigned int i = 1; i < (ARITY + 1); i++) {
+            handle_keys[i] = targets[i].c_str;
```

Comment:
```suggestion
            handle_keys[i] = targets[i-1].c_str;
```

Reply:
Ouch. It seems like my tests didn't covered these lines. Thanks!